rpc/src/fetch.rs

```diff
@@ -1,58 +1,64 @@
 use alloy_primitives::B256;
-/// Fetch data from related Portal networks
+use serde::{Deserialize, Serialize};
 use serde_json::Value;
 use tokio::sync::mpsc;
 
 use ethportal_api::{
     types::{
         execution::{block_body::BlockBody, header::Header},
-        jsonrpc::{endpoints::HistoryEndpoint, request::HistoryJsonRpcRequest},
+        history::ContentInfo,
+        jsonrpc::{
+            endpoints::HistoryEndpoint,
+            request::{HistoryJsonRpcRequest, JsonRpcRequest},
+        },
+        query_trace::QueryTrace,
     },
-    utils::bytes::hex_decode,
-    ContentValue, HistoryContentKey, HistoryContentValue,
+    HistoryContentKey, HistoryContentValue,
 };
 
-use crate::errors::RpcServeError;
+use crate::{errors::RpcServeError, serde::from_value};
 
-pub async fn proxy_query_to_history_subnet(
-    network: &mpsc::UnboundedSender<HistoryJsonRpcRequest>,
-    endpoint: HistoryEndpoint,
-) -> Result<Value, RpcServeError> {
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct ContentNotFoundError {
+    message: String,
+    trace: Option<QueryTrace>,
+}
+
+/// Fetch and deserialize data from Portal subnetwork.
+pub async fn proxy_to_subnet<TEndpoint, TOutput: serde::de::DeserializeOwned>(
+    network: &mpsc::UnboundedSender<JsonRpcRequest<TEndpoint>>,
+    endpoint: TEndpoint,
+) -> Result<TOutput, RpcServeError> {
     let (resp_tx, mut resp_rx) = mpsc::unbounded_channel::<Result<Value, String>>();
-    let message = HistoryJsonRpcRequest {
+    let message = JsonRpcRequest {
         endpoint,
         resp: resp_tx,
     };
     let _ = network.send(message);
 
-    match resp_rx.recv().await {
-        Some(val) => match val {
-            Ok(result) => Ok(result),
-            Err(msg) => {
-                if msg.contains("Unable to locate content on the network") {
-                    let error_details: Value = serde_json::from_str(&msg).map_err(|e| {
-                        RpcServeError::Message(format!(
-                            "Failed to parse error message from history subnet: {e:?}",
-                        ))
-                    })?;
-                    let message = error_details["message"]
-                        .as_str()
-                        .expect("Failed to parse error message from history subnet")
-                        .to_string();
-                    let trace = match error_details.get("trace") {
-                        Some(trace) => serde_json::from_value(trace.clone())
-                            .expect("Failed to parse query trace"),
-                        None => None,
-                    };
-                    Err(RpcServeError::ContentNotFound { message, trace })
-                } else {
-                    Err(RpcServeError::Message(msg))
-                }
+    let Some(response) = resp_rx.recv().await else {
+        return Err(RpcServeError::Message(
+            "Internal error: No response from subnetwork".to_string(),
```

Comment:
Can we also display the name of the subnetwork here and in the errors below? I think adding a trait to all subnetwork endpoints with some method that returns the name or implementing `DIsplay` should do the trick but maybe there is a simpler approach.